src/PanAndZoomContainer.js

```diff
@@ -59,17 +54,44 @@ const PanAndZoomContainer = ({
   );
 
   useEffect(() => {
-    diagramContainerRef.current.style.transform = transform;
-  }, [transform]);
-
-  useLayoutEffect(() => {
-    containerRef.current.scrollBy({
-      left:
-        (scroll.left * (zoom - previousZoom.current)) / previousZoom.current,
-      top: (scroll.top * (zoom - previousZoom.current)) / previousZoom.current
-    });
+    const container = containerRef.current;
+
+    const containerWidth = container.clientWidth;
+    const containerHeight = container.clientHeight;
+
+    let x = pan.x,
+      y = pan.y;
+
+    const extremeX = Math.max(contentSpan.x, containerWidth),
+      extremeY = Math.max(contentSpan.y, containerHeight);
+
+    if (containerWidth >= extremeX * zoom) {
+      x = ((containerWidth - extremeX) * zoom) / 2;
+    }
+    if (containerHeight >= extremeY * zoom) {
+      y = ((containerHeight - extremeY) * zoom) / 2;
+    }
+
+    diagramContainerRef.current.style.transform = `translate3D(${x}px, ${y}px, 0) scale(${zoom})`;
+
+    container.scrollLeft = containerRef.current.scrollLeft =
+      (scroll.left * zoom) / previousZoom.current +
+      ((zoom - previousZoom.current) * containerWidth) /
+        (2 * previousZoom.current);
+    container.scrollTop =
+      (scroll.top * zoom) / previousZoom.current +
+      (scroll.top
+        ? ((zoom - previousZoom.current) * containerHeight) /
```

Comment:
why do we need this conditional check?

Reply:
if scroll.top is zero, then we don't not to move vertical scroll.
For eg: If we are in 20% zoom, we dont want to scroll the diagram vertically unless user scroll a little bit.

Reply:
will that scenario also occur for `scroll.left` ?